### DISCUSSION TITLE
Multiple JK-BMS on a single ESP8266

### DISCUSSION THREAD

**Discussion by SergiCEES** [syssi/esphome-jk-bms#158] [Q&A] [answered]
Hi! A friend is trying to decide a setup for their 8 banks of 100ah 16s batteries (parallel), so there are going to be 8 BMS. He reached me out to help him to decide for a BMS and I wonder the following questions, if somebody could give as a hand here. :)

1. Is it possible to run a single (or a couple) of ESP8266 for all 8 JK-BMS?
2. If so, for redundancy and trying to avoid (if possible) Wi-Fi connection. It's possible to publish via mqtt or modbus wired directly on a Raspberry Pi (or the server he is going to have running HA) using ESP8266/ESP32 with Ethernet or USB.

Other option could be using a Victron Venus OS (on a raspberry or a Cerbo) and connect from the RS485 port to USB as it's done here: https://github.com/Louisvdw/dbus-serialbattery Maybe this is the simple one.

Thanks for your time, awesome project!

**Comment by syssi** [comment, marked as answer]
1. If you decide to use this project/implementation you should use an ESP32 because this microcontroller provides 3 hardware UARTs. You could try to assign additional devices using software serial but keep in mind the baudrate (115200 baud) is pretty high for software serial and could cause errors.

2. There are ESP32 boards out there providing an ethernet connection (also supported by ESPHome if you buy a support ETH chipset). If you want to interface the BMSs directly you should use the mpp-solar (a python implementation) or dbus-serialbattery project.

**Reply by syssi** [reply]
I will provide an example configuration tomorrow to demonstrate how to interface multiple BMS using a single ESP. You have to load the `uart`, the `jk_modbus` and the `jk_bms` component multiple times. Please keep in mind: The devices doesn't support modbus / device addresses. A UART interface / a serial connection per BMS is required.

**Reply by SergiCEES** [reply]
Thank you for your quick response, pretty logic. I will take a look at the other projects also. As he is in an early stage of his planning, we've time to check everything. Maybe the other solution would be using a Seplos BMS (i've been reading that they can be installed in series with can, and they trick the Venus Os saying that they're a pylontech battery, so de DVCC could work also. Keep going and congratulations on your project!

**Reply by syssi** [reply]
I've added an example configuration to interface two JK-BMS devices: https://github.com/syssi/esphome-jk-bms/blob/main/esp32-example-multiple-devices.yaml

**Reply by SergiCEES** [reply]
I will bookmark it! Thanks for your time and dedication!

**Reply by demoze** [reply]
it is possible by adding an analog switch or multiplexer  to your esp, For example with TI TMUX1309 you can have up to 4 serial line.   You only need to have some free output gpio to set the active serial before using it.

**Reply by syssi** [reply]
It's not supported out of the box. At the moment the component can be instantiated multiple times and assigned to different UARTs. Depending on the `update_interval` of each instance a request is sent and the response is received+decoded. In your scenario you have to orchestrate the requests+responses sequentially. After each request+response you have to change the multiplexer configuration to select the next BMS. This isn't possible without some restructuring of the code base.

**Reply by demoze** [reply]
yep

**Reply by 2Fblob** [reply]
I'm using BLE with one BMS at present.
Is it actually possible to run two BLE connections to connect two BMS? 
If not, is it possible to have one BLE and one Serial?

Thanks

### Update ###

For Anyone else needing, there is one such example here: [https://github.com/syssi/esphome-jk-bms/blob/main/esp32-example-multiple-devices.yaml](url)
It works ok!
